contracts/stellar-multicall/src/tests/test.rs

```diff
@@ -0,0 +1,302 @@
+#![cfg(test)]
+extern crate std;
+
+use soroban_sdk::testutils::Address as _;
+use soroban_sdk::{contract, contractimpl, vec, Address, Env, IntoVal, Val, Vec};
+use stellar_axelar_std::events::{fmt_last_emitted_event, Event};
+use stellar_axelar_std::interfaces::OwnableInterface;
+use stellar_axelar_std::{assert_contract_err, interfaces, mock_auth, IntoEvent, Ownable};
+
+use crate::error::ContractError;
+use crate::tests::bank_contract::{TestBankContract, TestBankContractClient};
+use crate::tests::subscriber_contract::{TestSubscriptionContract, TestSubscriptionContractClient};
+use crate::types::FunctionCall;
+use crate::{Multicall, MulticallClient};
+
+#[macro_export]
+macro_rules! construct_function_call {
+    ($contract:expr, $approver:expr, $client:ident . $function:ident ( $($arg:expr),* $(,)? )) => {{
+        FunctionCall {
+            contract: $contract.clone(),
+            approver: $approver.clone(),
+            function: soroban_sdk::Symbol::new(&$client.env, stringify!($function)),
+            args: vec![&$client.env, $($arg.into_val(&$client.env)),*],
+        }
+    }};
+}
+
+#[contract]
+#[derive(Ownable)]
+pub struct TestTarget;
+
+#[derive(Debug, PartialEq, Eq, IntoEvent)]
+pub struct ExecutedEvent {
+    pub value: u32,
+}
+
+#[contractimpl]
+impl TestTarget {
+    pub fn __constructor(env: Env, owner: Address) {
+        interfaces::set_owner(&env, &owner);
+    }
+
+    pub fn method(env: &Env, value: u32) {
+        ExecutedEvent { value }.emit(env);
+    }
+
+    pub fn failing(_env: &Env) {
+        panic!("This method should fail");
+    }
+
+    pub const fn failing_with_error(_env: &Env) -> Result<Val, ContractError> {
+        Err(ContractError::FunctionCallFailed)
+    }
+}
+
+pub struct TestConfig<'a> {
+    pub env: Env,
+    pub client: MulticallClient<'a>,
+    pub target_id: Address,
+    pub owner: Address,
+}
+
+fn setup<'a>() -> TestConfig<'a> {
+    let env = Env::default();
+    let owner = Address::generate(&env);
+    let contract_id = env.register(Multicall, ());
+    let client = MulticallClient::new(&env, &contract_id);
+
+    let target_id = env.register(TestTarget, (owner.clone(),));
+
+    TestConfig {
+        env,
+        client,
+        target_id,
+        owner,
+    }
+}
+
+#[test]
+fn multicall_succeeds() {
+    let TestConfig {
+        env,
+        client,
+        target_id,
+        owner,
+    } = setup();
+
+    let bank_id = env.register(TestBankContract, (owner.clone(),));
```

Comment:
if it's small enough, this contract could be defined in this file since it's not used anywhere else